Source/Diagnostics/WarpXOpenPMD.cpp

```diff
@@ -0,0 +1,479 @@
+
+{
+  int numOutputReal = 0;
+  int totalRealAttrs = m_NumAoSRealAttributes + m_NumSoARealAttributes;
+
+  for (int i = 0; i < totalRealAttrs; ++i)
+    if (write_real_comp[i])
+      ++numOutputReal;
+
+  auto numParticleOnTile = pti.numParticles();
+  const auto& aos = pti.GetArrayOfStructs();  // size =  numParticlesOnTile
+  const auto& soa = pti.GetStructOfArrays();
+
+  // properties are saved seperately
+  {
+    for (auto idx=0; idx<m_NumAoSRealAttributes; idx++) {
+      if (write_real_comp[idx]) {
+          auto& currVar = currSpecies[real_comp_names[idx]][openPMD::RecordComponent::SCALAR];
+          typename amrex::ParticleReal *d =
+                 static_cast<typename amrex::ParticleReal*> (malloc(sizeof(typename amrex::ParticleReal) *  numParticleOnTile));
+
+          for (auto kk=0; kk<numParticleOnTile; kk++)
+               d[kk] = aos[kk].m_rdata.arr[AMREX_SPACEDIM+idx];
+
+          std::shared_ptr <typename amrex::ParticleReal> data(d, free);
+          currVar.storeChunk(data,
+               {offset}, {static_cast<unsigned long long>(numParticleOnTile)});
+          m_Series->flush();
+      }
+    }
+  }
+
+  {
+    for (auto idx=0; idx<m_NumSoARealAttributes; idx++) {
+      auto ii = m_NumAoSRealAttributes + idx;
+      if (write_real_comp[ii]) {
+          auto& currVar = currSpecies[real_comp_names[ii]][openPMD::RecordComponent::SCALAR];
+          currVar.storeChunk(openPMD::shareRaw(soa.GetRealData(idx)),
+                             {offset}, {static_cast<unsigned long long>(numParticleOnTile)});
+      }
+    }
+    m_Series->flush();
+  }
+}
+
+
+
+void
+WarpXOpenPMDPlot::SetupPos(openPMD::ParticleSpecies& currSpecies,
+               const unsigned long long& np) const
+{
+  auto particleLineup = openPMD::Dataset(openPMD::determineDatatype<amrex::ParticleReal>(), {np});
+
+  currSpecies["positionOffset"]["x"].resetDataset(particleLineup);
+  currSpecies["positionOffset"]["x"].makeConstant(0);
+  currSpecies["positionOffset"]["y"].resetDataset(particleLineup);
+  currSpecies["positionOffset"]["y"].makeConstant(0);
+  currSpecies["positionOffset"]["z"].resetDataset(particleLineup);
+  currSpecies["positionOffset"]["z"].makeConstant(0);
+
+  //auto positions = openPMD::Dataset(openPMD::determineDatatype<amrex::ParticleReal>(), {np});
+  currSpecies["position"]["x"].resetDataset(particleLineup);
+  currSpecies["position"]["y"].resetDataset(particleLineup);
+  currSpecies["position"]["z"].resetDataset(particleLineup);
+}
+
+
+
+
+//
+// this is originally copied from FieldIO.cpp
+//
+void
+WarpXOpenPMDPlot::WriteOpenPMDFields( //const std::string& filename,
+                      const std::vector<std::string>& varnames,
+                      const amrex::MultiFab& mf,
+                      const amrex::Geometry& geom,
+                      const int iteration,
+                      const double time ) const
+{
+  BL_PROFILE("WriteOpenPMDFields()");
```

Comment:
Since you added `Timer` functionality: This should already add timings to the amrex "tiny profiler" timer, just in case.